tests/ludwig/models/test_training_determinism.py

```diff
@@ -0,0 +1,109 @@
+import json
+import logging
+import os
+
+import pytest
+
+from ludwig.api import LudwigModel
+from ludwig.constants import TRAINER
+from ludwig.utils.data_utils import NumpyEncoder
+from tests.integration_tests.utils import (
+    audio_feature,
+    bag_feature,
+    binary_feature,
+    category_feature,
+    date_feature,
+    generate_data,
+    h3_feature,
+    image_feature,
+    number_feature,
+    sequence_feature,
+    set_feature,
+    text_feature,
+    timeseries_feature,
+    vector_feature,
+)
+
+
+@pytest.mark.distributed
+def test_training_determinism_ray_backend(csv_filename, tmpdir):
+    experiment_output_1, experiment_output_2 = train_twice("ray", csv_filename, tmpdir)
+
+    eval_stats_1, train_stats_1, _, _ = experiment_output_1
+    eval_stats_2, train_stats_2, _, _ = experiment_output_2
+
+    assert json.dumps(eval_stats_1, cls=NumpyEncoder, sort_keys=True, indent=4) == json.dumps(
+        eval_stats_2, cls=NumpyEncoder, sort_keys=True, indent=4
+    )
+    assert json.dumps(train_stats_1, cls=NumpyEncoder, sort_keys=True, indent=4) == json.dumps(
```

Comment:
Are the training stats won't always be the exact same, even with full determinism? I'd assume that there's some decimal errors, in which case doing an `np.allclose` might be more appropriate instead of an `equals` assertion.

Reply:
Used `numpy.testing.assert_equal` which consider `np.nan` as if they were normal numbers.